### ISSUE TITLE
Add UDP

### ISSUE THREAD

**Issue by aratz-lasa** [libp2p/py-libp2p#326]
I have made a first draft for implementing UDP in Transport Layer.
It is supposed to be implemented in Go, however master branch is [empty](https://github.com/libp2p/go-udp-transport/). The only code we can find is in the second branch [feat/udp](https://github.com/libp2p/go-udp-transport/tree/feat/udp).

So, my approach is not based at all in Go's "implementation". I propose a unified interface for UDP and TCP Streams: [IStreamReader](https://github.com/aratz-lasa/py-libp2p/blob/5191704ff34ad9aef48f94e4b380d03ad1e5325e/libp2p/transport/stream_interface.py#L15) and [IStreamWriter](https://github.com/aratz-lasa/py-libp2p/blob/5191704ff34ad9aef48f94e4b380d03ad1e5325e/libp2p/transport/stream_interface.py#L21). Apart from getting a unified interface, I have also tried to keep it simple. Now, streams just implement four methods: **get_extra_info**, **close**, **read** and **write**. I may have made it too high-level abstracting from methods like **drain** or others. I let it for discussion ;) 

 It may look weird UDP implementing a Stream interface, however, in order to achieve modularity between transports, we must simulate a Session-like-stream on UDP. Due to this Unified Interface, every transport protocol will need to implement four interfaces: **ITransport**, **IListener** and **IStreamReader** and **IStreamWriter**. You can find the new Stream interfaces here: [UDP Streams](https://github.com/aratz-lasa/py-libp2p/blob/add-udp/libp2p/transport/udp/udp_stream.py) and [TCP Streams](https://github.com/aratz-lasa/py-libp2p/blob/add-udp/libp2p/transport/tcp/tcp_stream.py). On the other hand, I think this two last interfaces could be gatherer into one interface, however, I have prefered to divide them, for being compatible with current **RawConnection**, **Swarm**....
  
Of course, this code is just a playeable implementation, intended for encouraging discussions and give a new perspective. I have not made any tests for UDP yet, even though, TCP passes current tests while implementing this new Unified Interface.

**Comment by ralexstokes** [comment]
i know there has been some discussion on how to best unify tcp and udp under libp2p -- pinging @mgoelzer or @raulk for links to previous discussion

**Comment by mhchia** [comment]
1. Unified interface: to me, the unified interface for different transport sounds nice and necessary, in order to support other transports like QUIC.

2. UDP: I couldn't find out a component in libp2p where UDP is necessarily depended on in other implementations. What I can think of for now is Kademlia, but the currently used transport which IPFS run Kademlia against is TCP. I lean toward to have more discussion on its usage and how we should support it before we actually implement UDP.

**Comment by aratz-lasa** [comment]
>     2. UDP: I couldn't find out a component in libp2p where UDP is necessarily depended on in other implementations. What I can think of for now is Kademlia, but the currently used transport which IPFS run Kademlia against is TCP. I lean toward to have more discussion on its usage and how we should support it before we actually implement UDP.

I completely agree. We must think and discuss about it, before implementing it. (By the way, I just did the pull request for being easily referable the code, I did not have any intention for merging it right away)

What do you mean by "how we should support it"? How would be best implemented?

**Comment by mhchia** [comment]
I had a post asking about [UDP support](https://discuss.libp2p.io/t/is-there-any-discussion-on-how-we-should-unify-the-interface-for-tcp-and-udp-transports/319/5) in libp2p forum. Since the discussion on the packet protocol is still ongoing, IMO it's nice to contribute to the proposal and we can implement it after things are settled.

**Comment by mgoelzer** [comment]
@bigs

**Comment by raulk** [comment]
Hey folks, track and participate in the discussion by following the links on this epic: https://github.com/libp2p/specs/issues/225 Thanks!

**Comment by ShadowJonathan** [comment]
Maybe it's interesting to note here that I have been thinking about how to reorganize libp2p's connection implimentation structure, for example, [here](https://files.gitter.im/py-libp2p/Lobby/FXsZ/First-version.png) is a quick diagram of how the current tree (partially) looks like in that regard, and I've made a theoretical implimentation diagram [here](https://i.imgur.com/KVbKXpj.png) of a new situation, this all ties into #228.

The last diagram is still a draft, one which I have to flesh out and add other parts of the program to to make an actual calculated proposal in an issue I'll post here.

(Opinion: Python is not Golang, and we shouldn't entirely rely on go-libp2p for examples, but instead work and live by the specs)

**Comment by pacrob** [comment]
We've decided UDP is no longer necessary, QUIC is now the standard.
